File: 129-sum-root-to-leaf-numbers/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        # return self.dfs(root, 0)
        return self.iterative(root)
    
    # def dfs(self, root: Optional[TreeNode], sum_v: int) -> int:
    #     if root is None:
    #         return 0
    #     sum_v = sum_v * 10 + root.val
    #     if (root.left is None and root.right is None):
    #         return sum_v
    #     return self.dfs(root.left, sum_v) + self.dfs(root.right, sum_v)

    def iterative(self, root):
        res = 0
        stack = [root]
        while len(stack) != 0:
            node = stack.pop()
            if node.left is None and node.right is None:
                res += node.val
            
            if node.left is not None:
                node.left.val += node.val * 10
                stack.append(node.left)
            
            if node.right is not None:
                node.right.val += node.val * 10
                stack.append(node.right)
        return res

File: 129-sum-root-to-leaf-numbers/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        if (!root) return 0;
        int res = 0;
        stack<TreeNode*> st;
        st.push(root);
        while (st.size() > 0) {
            auto node = st.top(); st.pop();
            if (!node->left && !node->right) {
                res += node->val;
            }

            if (node->left) {
                node->left->val += node->val * 10;
                st.push(node->left); 
            }
            if (node->right) {
                node->right->val += node->val * 10;
                st.push(node->right); 
            }
        }
        return res;
    }
};
